File: androidutil/src/main/java/com/akingyin/base/time/AppTime.java
/*
 * Copyright (c) 2017. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
 * Morbi non lorem porttitor neque feugiat blandit. Ut vitae ipsum eget quam lacinia accumsan.
 * Etiam sed turpis ac ipsum condimentum fringilla. Maecenas magna.
 * Proin dapibus sapien vel ante. Aliquam erat volutpat. Pellentesque sagittis ligula eget metus.
 * Vestibulum commodo. Ut rhoncus gravida arcu.
 */

package com.akingyin.base.time;

import android.os.SystemClock;

import com.akingyin.base.user.UserInfoEntity;
import com.akingyin.base.user.UserSingleInfo;


/**
 * 终端时间处理
 * @ Description:
 * @author king
 * @ date 2017/9/13 13:06
 * @version  V1.0
 */
public class AppTime {
  private volatile static AppTime singleton;

  private   Long   loginTme;

  private   Long   elapsedRealtime;

  public void setLoginTme(Long loginTme) {
    this.loginTme = loginTme;
  }

  public void setElapsedRealtime(Long elapsedRealtime) {
    this.elapsedRealtime = elapsedRealtime;
  }

  public  synchronized   Long    getTime(){
    if(null == loginTme || null == elapsedRealtime){
      UserInfoEntity user = UserSingleInfo.getInstance().get();
      if(null != user && null != user.loginTime && null != user.elapsedRealtime){
        loginTme = user.loginTime;
        elapsedRealtime = user.elapsedRealtime;
      }else{
        return  System.currentTimeMillis();
      }
    }
    return  loginTme+(SystemClock.elapsedRealtime() - elapsedRealtime);
  }

  public static AppTime getInstance() {
    if (singleton == null) {
      synchronized (AppTime.class) {
        if (singleton == null) {
          singleton = new AppTime();
        }
      }
    }
    return singleton;
  }

  private AppTime() {
  }
}

File: androidutil/src/main/java/com/akingyin/base/ext/AppTime.kt
package com.akingyin.base.ext

import android.os.SystemClock

/**
 * @ Description:
 * @author king
 * @ Date 2019/5/28 11:21
 * @version V1.0
 */
class AppTime  private constructor(){

     /** 服务器时间 */
     private   var   serverTime : Long = 0L

     /** 手机运行时间 */
     private   var   elapsedRealtime : Long = 0L

     /** 登录超期时间 */
     private   var   extendedTime:Long = 0L

    companion object{
         private val Instance: AppTime by lazy { AppTime() }


        /**
         * 获取当前时间
         */
        fun   getAppTime():Long{
             if(Instance.serverTime == 0L){
                 Instance.serverTime = spGetLong("serverTime",0L)
             }
             if(Instance.elapsedRealtime == 0L){
                 Instance.serverTime = spGetLong("elapsedRealtime",0L)
             }
             if(Instance.serverTime == 0L || Instance.elapsedRealtime == 0L){
                 return  System.currentTimeMillis()
             }


             return  Instance.serverTime + (SystemClock.elapsedRealtime() - Instance.elapsedRealtime)
         }

         fun   saveServerTime(server:Long){
             Instance.serverTime = server
             Instance.elapsedRealtime = SystemClock.elapsedRealtime()
             spSetLong("serverTime",server)
             spSetLong("elapsedRealtime", Instance.elapsedRealtime)
         }

         fun   saveExtendedTime(extendedTime:Long){
              Instance.extendedTime = extendedTime
              spSetLong("extendedTime",extendedTime)
         }

         fun  getExtendedTime():Long{
             if(Instance.extendedTime == 0L){
                 Instance.extendedTime = spGetLong("extendedTime")
             }
             return  Instance.extendedTime
         }

    }


}